tcf_website/views/browse.py

```diff
@@ -75,36 +75,26 @@ def course_view(request, course_id):
 
     course = Course.objects.get(pk=course_id)
     latest_semester = Semester.latest()
-
-    # Get instructors that have taught the course in the most recent
-    # semester.
-    recent_instructor_pks = course.section_set.filter(
-        semester=latest_semester).values_list(
-        'instructors', flat=True).distinct()
-    recent_instructors = Instructor.objects.filter(
-        pk__in=recent_instructor_pks)
-    # Add ratings and difficulties
-    for instr in recent_instructors:
-        instr.rating = instr.average_rating_for_course(course)
-        instr.difficulty = instr.average_difficulty_for_course(course)
-        instr.gpa = CourseInstructorGrade.objects.filter(
-            course=course, instructor=instr).aggregate(
-            Avg('average')).get('average__avg')
-
-    # Get instructors that haven't taught the course this semester.
-    old_instructor_pks = course.section_set.exclude(
-        semester=latest_semester).exclude(
-        instructors__pk__in=recent_instructor_pks).values_list(
-        'instructors',
-        flat=True).distinct()
-    old_instructors = Instructor.objects.filter(pk__in=old_instructor_pks)
-    # Add ratings and difficulties
-    for instr in old_instructors:
-        instr.rating = instr.average_rating_for_course(course)
-        instr.difficulty = instr.average_difficulty_for_course(course)
-        instr.gpa = CourseInstructorGrade.objects.filter(
-            course=course, instructor=instr).aggregate(
-            Avg('average'))['average__avg']
+    instructors = Instructor.objects\
+        .filter(section__course=course).distinct()\
+        .annotate(
+            gpa=Avg('courseinstructorgrade__average',
+                    filter=Q(courseinstructorgrade__course=course)),
+            difficulty=Avg('review__difficulty', filter=Q(review__course=course)),
+            rating=(
+                Avg('review__instructor_rating', filter=Q(review__course=course)) +
+                Avg('review__enjoyability', filter=Q(review__course=course)) +
+                Avg('review__recommendability', filter=Q(review__course=course))
+            ) / 3,
+            semester_last_taught=Max('section__semester',
+                                     filter=Q(section__course=course)),
+        )
+    # Note: Wanted to use .annotate() but couldn't figure out a way
+    # So created a dictionary on the fly to minimize database access
+    semesters = {s.id: s for s in Semester.objects.all()}
+    for instructor in instructors:
+        instructor.semester_last_taught = semesters.get(
+            instructor.semester_last_taught)
```

Comment:
Instead of iterating through the loop to check against `latest_semester` (which, to my understanding, is currently an instance of Semester), wouldn't it be easier to make `latest_semester` also just an ID to check against?

Reply:
`latest_semester` is used in the template https://github.com/thecourseforum/theCourseForum2/blob/a913e825459c23b09df9daa5d452a3aa6d6910dc/tcf_website/templates/course/course.html#L18 so we need to access the `Semester` table anyway.

Reply:
It might be better to compare `i.semester_last_taught` against `lastest_semester.id` (not sure if that's the right way to do it) in the template then?

Reply:
update: can confirm you can access the id with `latest_semester.id` — definitely think that's the way to go instead of changing fields

Reply:
Okay, let's start from the basics to make sure we're on the same page.
1. Before this part of the code, I have as `semester_last_taught=Max('section__semester', filter=Q(section__course=course))`, which is an `int`. (I couldn't figure out a way to make it an instance of `Semester`, and I don't think it's possible.)
2. However, what the template needs is an instance of `Semester` (or a string representation thereof) for each `Instructor`. That's why I need to access the database again. However, accessing the database within a loop would be very inefficient, so instead I created a `Dict[int, Semester]` in line 94.
3. Going back to your first comment, I'm not sure why you brought up `latest_semester`, which is the most recent semester at the moment. What we need is the most recent semester *for each `Instructor` teaching a given course*, which I denote as `semester_last_taught`.
4. As I said in 1, I only have the ID (which is an `int`) of the semester last taught. Line 96 is converting the `int` to an instance of `Semester` for the reason mentioned in 2, which is possible because Python is dynamically typed.
5. Just so we're on the same page, the conversion in 4 does change a field of an `Instructor` instance, but we didn't call `.save()`, so the change does not propagate to the database (which should fail anyway).

With that said, can you clarify why `latest_semester` was mentioned?

Reply:
Thanks for the clarification — I overlooked one of the sections of the code where you pass the `Semester` instance itself into the `rating_card.html` template... I saw the `{% if i.semester_last_taught != latest_semester %}` refactor and assumed that this could be simplified with just using IDs. Sorry for the confusion!